### PR TITLE
Manage xacml

### PR DESCRIPTION
**Pull request by MorganDawe**
Allows for one to choose when to apply XACML policy's to a child collection and which parent to inherit from. This can be done pre/post ingest.

### CODE REVIEW

File: islandora_xacml_editor.module [hunk 1]
```diff
@@ -252,3 +255,171 @@ function islandora_xacml_editor_undo_policy(&$form_state) {
...
+/**
+ * Form alter the child collection form.
+ *
+ * @param array $form
+ *   The form being passed from the hook.
+ * @param array $form_state
+ *   The current Form State being processed.
+ */
+function islandora_xacml_editor_form_islandora_basic_collection_create_child_collection_form_alter(array &$form, array &$form_state) {
+  dsm($form);
```
Comment (mitchmac) [inline islandora_xacml_editor.module]: dsm is bad, mmkay

File: islandora_xacml_editor.module [hunk 2]
```diff
@@ -252,3 +255,171 @@ function islandora_xacml_editor_undo_policy(&$form_state) {
...
+  $parent_object = islandora_object_load($form_state['islandora']['shared_storage']['parent']);
+  $xacml_options = array('None' => 'none');
+  if ($parent_object) {
+    $xacml_options[$parent_object->id] = $parent_object->label;
+  }
+
+  $form['xacml'] = array(
+    '#weight' => 1,
```
Comment (mitchmac) [inline islandora_xacml_editor.module]: Should try to set this dynamically, to be 'lighter' than the form submit button.

File: islandora_xacml_editor.module [hunk 3]
```diff
@@ -252,3 +255,171 @@ function islandora_xacml_editor_undo_policy(&$form_state) {
...
+ *
+ * @return array
+ *   The Drupal form definition.
+ */
+function islandora_xacml_editor_manage_xacml_form(array $form, array &$form_state, AbstractObject $object) {
+  module_load_include('inc', 'islandora', 'includes/utilities');
+  $child_collections_pids = islandora_basic_collection_get_child_pids($object, 'islandora:collectionCModel');
+  $rows = array();
+  $options = array('none' => 'None');
+  $child_collection;
```
Comment (mitchmac) [inline islandora_xacml_editor.module]: Probably don't want this.

File: islandora_xacml_editor.module [hunk 4]
```diff
@@ -252,3 +255,171 @@ function islandora_xacml_editor_undo_policy(&$form_state) {
...
+/**
+ * Submit handler for the manage XACML form.
+ *
+ * @param array $form
+ *   The Drupal form definition.
+ * @param array $form_state
+ *   The Drupal form state.
+ */
+function islandora_xacml_editor_manage_xacml_form_submit(array $form, array &$form_state) {
+  $parent_object;
```
Comment (mitchmac) [inline islandora_xacml_editor.module]: Probably don't want this.

File: islandora_xacml_editor.module [hunk 5]
```diff
@@ -252,3 +255,171 @@ function islandora_xacml_editor_undo_policy(&$form_state) {
...
+ *   The Drupal form state.
+ * @param AbstractObject $object
+ *   The collection to move child objects from.
+ *
+ * @return array
+ *   The Drupal form definition.
+ */
+function islandora_xacml_editor_manage_xacml_form(array $form, array &$form_state, AbstractObject $object) {
+  module_load_include('inc', 'islandora', 'includes/utilities');
+  $child_collections_pids = islandora_basic_collection_get_child_pids($object, 'islandora:collectionCModel');
```
Comment (mitchmac) [inline islandora_xacml_editor.module]: Looks like this is deprecated in favor of islandora_basic_collection_get_member_objects

File: islandora_xacml_editor.module [hunk 6]
```diff
@@ -252,3 +255,171 @@ function islandora_xacml_editor_undo_policy(&$form_state) {
...
+    'help' => array(
+      '#type' => 'item',
+      '#markup' => t('XACML Inheritance Policies'),
+    ),
+    'table' => array(
+      '#tree' => TRUE,
+      '#header' => array(
+        array('class' => array('select-all')), t('COLLECTION(PID)'), 'INHERIT FROM',
+      ),
+      '#theme' => 'islandora_basic_collection_policy_management_table',
```
Comment (mitchmac) [inline islandora_xacml_editor.module]: Using this sets the empty table message to "No content models" which is confusing.

File: islandora_xacml_editor.module [hunk 7]
```diff
@@ -216,8 +216,11 @@ function islandora_xacml_editor_islandora_ingest_steps(array $form_state) {
  */
 function islandora_xacml_editor_apply_policy(&$form_state, $parent) {
   $parent_object = islandora_object_load($parent);
-  if ($parent_object['POLICY']) {
+  if ($parent_object['POLICY'] && $form_state['input']['xacml'] != 'None') {
```
Comment (mitchmac) [inline islandora_xacml_editor.module]: I think we should throw a message if the parent policy isn't set, I was derping wondering why nothing was happening.

File: islandora_xacml_editor.module [hunk 8]
```diff
@@ -216,8 +216,11 @@ function islandora_xacml_editor_islandora_ingest_steps(array $form_state) {
...
     foreach ($form_state['islandora']['objects'] as $object) {
+      // Need to update the RELS-INT to indicate which parent this policy is
+      // inherited from.
+      $object->relationships->add(FEDORA_RELS_EXT_URI, 'inheritXacmlFrom', $parent, TRUE);
```
Comment (mitchmac) [inline islandora_xacml_editor.module]: I think we should keep this in the Islandora predicate ns/uri

File: islandora_xacml_editor.module [hunk 9]
```diff
@@ -252,3 +269,194 @@ function islandora_xacml_editor_undo_policy(&$form_state) {
...
+            $object[$dsid]->relationships->remove(ISLANDORA_RELS_INT_URI, $viewable_by_user);
+            $object[$dsid]->relationships->remove(ISLANDORA_RELS_INT_URI, $viewable_by_role);
+            $object[$dsid]->relationships->remove(ISLANDORA_RELS_INT_URI, $manageable_by_user);
+            $object[$dsid]->relationships->remove(ISLANDORA_RELS_INT_URI, $manageable_by_role);
+          }
+        }
+      }
+    }
+    $count++;
+  }
```
Comment (mitchmac) [inline islandora_xacml_editor.module:404]: Should set a message if something happens with the submit.

File: islandora_xacml_editor.module [hunk 10]
```diff
@@ -252,3 +269,207 @@ function islandora_xacml_editor_undo_policy(&$form_state) {
...
+ *   The Islandora Object being processed.
+ *
+ * @return array
+ *   The current Form State with manage_xacml element appended.
+ */
+function islandora_xacml_editor_islandora_basic_collection_build_manage_object(array $form_state, AbstractObject $object) {
+  $form_state['manage_collection_object']['manage_xacml'] = array(
+    '#id' => 'manage-xacml',
+    '#group' => 'manage_xacml_object',
+    '#access' => TRUE,
```
Comment (mitchmac) [inline islandora_xacml_editor.module]: Seems there are no permissions to check? Shouldn't be wide open.

File: islandora_xacml_editor.module (hunk 9 above)
Comment (MorganDawe) [inline islandora_xacml_editor.module:404]: Latest push has a drupal_set_message() call after the for loop

File: islandora_xacml_editor.module [hunk 11]
```diff
@@ -4,6 +4,9 @@
  * The main module file for the Islanora XACML Editor.
  */
 
+// Permission.
+define('ISLANDORA_XACML_ADMINISTER_INHERITANCE_PERMISSION', 'administer xacml inheritance');
```
Comment (willtp87) [inline islandora_xacml_editor.module]: Constant not prefixed with module name.

File: islandora_xacml_editor.module [hunk 12]
```diff
@@ -252,3 +275,198 @@ function islandora_xacml_editor_undo_policy(&$form_state) {
...
+          $xacml->writeBackToFedora();
+          $message .= $object->id . " now inherits XACML from " . $parent_object->id . "\n";
+        }
+      }
+      else {
+        // 'None' is currently selected for this row, so if it is selected,
+        // we must remove the current XACML policy if it exists.
+        $object = islandora_object_load($child_pids[$count]);
+        if (isset($object['POLICY'])) {
+          $viewable_by_user = 'isViewableByUser';
```
Comment (willtp87) [inline islandora_xacml_editor.module]: This looks like a hole in the API, we should move this code lower for re-usability.

File: islandora_xacml_editor.module [hunk 13]
```diff
@@ -252,3 +213,183 @@ function islandora_xacml_editor_undo_policy(&$form_state) {
...
+      // Are we adding the XACML POLICY or removing it? Just
+      // check for a valid fedora object.
+      if ($parent_object) {
+        // Make sure there is a POLICY to inherit.
+        if ($parent_object['POLICY']) {
+          $object = islandora_object_load($child_pids[$count]);
+          $object->relationships->add(FEDORA_RELS_EXT_URI, 'inheritXacmlFrom', $parent_object, TRUE);
+          $xacml = new IslandoraXacml($object, $parent_object['POLICY']->content);
+          $xacml->writeBackToFedora();
+          $message .= $object->id . " now inherits XACML from " . $parent_object->id . "\n";
```
Comment (mitchmac) [inline islandora_xacml_editor.module]: Messages can be set separately with their own drupal_set_message. Then they need proper t() use with placeholders for variables.

File: islandora_xacml_editor.module [hunk 14]
```diff
@@ -252,3 +213,190 @@ function islandora_xacml_editor_undo_policy(&$form_state) {
...
+          $object->relationships->add(FEDORA_RELS_EXT_URI, 'inheritXacmlFrom', $parent_object, TRUE);
+          $xacml = new IslandoraXacml($object, $parent_object['POLICY']->content);
+          $xacml->writeBackToFedora();
+          drupal_set_message(t('@child now inherits XACML from @parent.',
+            array(
+              '@child' => $object->id,
+              '@parent' => $parent_object->id),
+            'status')
+          );
+        }
```
Comment (mitchmac) [inline islandora_xacml_editor.module:388]: If the parent doesn't have a policy, it's confusing why nothing is happening. Should set a message similar to the other case where the parent doesn't have a policy.

File: islandora_xacml_editor.module [hunk 15]
```diff
@@ -252,3 +213,190 @@ function islandora_xacml_editor_undo_policy(&$form_state) {
...
+        // Make sure there is a POLICY to inherit.
+        if ($parent_object['POLICY']) {
+          $object = islandora_object_load($child_pids[$count]);
+          $object->relationships->add(FEDORA_RELS_EXT_URI, 'inheritXacmlFrom', $parent_object, TRUE);
+          $xacml = new IslandoraXacml($object, $parent_object['POLICY']->content);
+          $xacml->writeBackToFedora();
+          drupal_set_message(t('@child now inherits XACML from @parent.',
+            array(
+              '@child' => $object->id,
+              '@parent' => $parent_object->id),
```
Comment (mitchmac) [inline islandora_xacml_editor.module]: There's a syntax error here, t() should be closed after this array.

File: islandora_xacml_editor.module [hunk 16]
```diff
@@ -252,3 +213,190 @@ function islandora_xacml_editor_undo_policy(&$form_state) {
...
+      }
+      else {
+        // 'None' is currently selected for this row, so if it is selected,
+        // we must remove the current XACML policy if it exists.
+        $object = islandora_object_load($child_pids[$count]);
+        if (isset($object['POLICY'])) {
+          $object->relationships->remove(FEDORA_RELS_EXT_URI, 'inheritXacmlFrom', $parent_object, TRUE);
+          $object->purgeDatastream('POLICY');
+          drupal_set_message(t('@child no longer inherits XACML.',
+            array('@object' => $object->id),
```
Comment (mitchmac) [inline islandora_xacml_editor.module]: There's a syntax error here, t() should be closed after this array.

File: islandora_xacml_editor.module [hunk 17]
```diff
@@ -252,3 +213,194 @@ function islandora_xacml_editor_undo_policy(&$form_state) {
...
+      }
+      else {
+        // 'None' is currently selected for this row, so if it is selected,
+        // we must remove the current XACML policy if it exists.
+        $object = islandora_object_load($child_pids[$count]);
+        if (isset($object['POLICY'])) {
+          $object->relationships->remove(FEDORA_RELS_EXT_URI, 'inheritXacmlFrom', $parent_object, TRUE);
+          $object->purgeDatastream('POLICY');
+          drupal_set_message(t('@child no longer inherits XACML.',
+            array('@object' => $object->id)),
```
Comment (mitchmac) [inline islandora_xacml_editor.module]: Need the placeholder value for child

File: islandora_xacml_editor.module [hunk 18]
```diff
@@ -252,3 +213,194 @@ function islandora_xacml_editor_undo_policy(&$form_state) {
...
+ *
+ * @param array $form
+ *   The Drupal form definition.
+ * @param array $form_state
+ *   The Drupal form state.
+ */
+function islandora_xacml_editor_manage_xacml_form_submit(array $form, array &$form_state) {
+  $child_pids = array_keys($form_state['values']['table']['rows']);
+  $count = 0;
+  $message = "Complete: \n";
```
Comment (mitchmac) [inline islandora_xacml_editor.module]: Unused variable.

File: islandora_xacml_editor.module [hunk 19]
```diff
@@ -252,3 +213,194 @@ function islandora_xacml_editor_undo_policy(&$form_state) {
...
+          $xacml->writeBackToFedora();
+          drupal_set_message(t('@parent does not have an XACML policy.',
+            array(
+              '@parent' => $parent_object->id)),
+            'status');
+        }
+        else {
+          drupal_set_message(t('@child now inherits XACML from @parent.',
+            array(
+              '@child' => $object->id,
```
Comment (mitchmac) [inline islandora_xacml_editor.module]: $object isn't around here, is this message reversed with the one above it?

**Pull request merged**